### PR TITLE
tests/multi_net/udp_data.py: Make UDP test more reliable.

### PR DESCRIPTION
**Pull request by iabdalkader**
The current test depends on a specific number and order of packets to pass, which can't be reproduced every run due to the unreliable UDP protocol. This patch adds simple packets sequencing, retransmits with timeouts, and a packet loss threshold, to make the test more tolerant to UDP protocol packet drops and reordering.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Is there some deterministic printing that can be done by both client and server, so that you can see that progress is being made during the test?  When it doesn't work and you use `-t` to trace the output it's helpful to know if it's just slow, or if it cannot reach the device at all.

Maybe print the sequence number until it hits the threshold of success?

**Comment by iabdalkader** [comment]
> Maybe print the sequence number until it hits the threshold of success?

Yes, I also realized I shouldn't `break` in inner loops because I was basically sending maximum of 4 x 1 packets.

**Comment by dpgeorge** [comment]
I would have thought a way to make it reliable would be:
- client keeps retrying to send "client send %d" (with max num retries) until it gets a response "server reply %d"
- client does that in a loop, eg 10 times (incrementing the counter for %d)
- client prints when it gets a valid response
- server just responds to %d with same %d
- server prints when it gets a new incoming message

**Comment by iabdalkader** [comment]
Fixed and printing sequence as long as it's less than loss threshold.

**Comment by dpgeorge** [comment]
Ok, you've basically done what I wrote above :smile:

**Comment by iabdalkader** [comment]
> Ok, you've basically done what I wrote above smile

Yes :) side note though, the timeout (10s ?) in `./run-multitests.py` puts an upper limit on the number of packets/retries given that each one could take up to ~250ms, and that the test stops printing after 60% of successful packets, it's possible that the remaining packets cause the test not to output anything for a long time, depending on the number of remaining packets. This is Not possible with the default 4*10 packets.

**Comment by iabdalkader** [comment]
Update: simplified the test.

File: tests/multi_net/udp_data.py [hunk 1]
```diff
@@ -3,34 +3,69 @@
...
+                s.sendto(b"%d" % (seq), addr)
         s.close()
 
+    if (TOTAL_PACKETS - seq) > PACKET_LOSS_THRESH:
+        print(
+            "packet loss %.1f%% %d/%d"
+            % (((TOTAL_PACKETS - seq) / TOTAL_PACKETS * 100), seq, TOTAL_PACKETS)
+        )
+    else:
+        print("pass")
```
Comment (dpgeorge) [inline tests/multi_net/udp_data.py:41]: This bit of code seems common to instance0 and instance1, so best if it's made into a function they can both call.

Comment (iabdalkader) [reply tests/multi_net/udp_data.py:43]: Fixed

Comment (dpgeorge) [reply tests/multi_net/udp_data.py:43]: looks like this changed was not pushed here??

Comment (iabdalkader) [reply tests/multi_net/udp_data.py:43]: It did, I moved the code that was here to `loss_percent()`. What else do you want to make common ?

Comment (dpgeorge) [reply tests/multi_net/udp_data.py:43]: I thought these 4 lines (the whole if-else) can be made a function, with argument `seq`.

Comment (dpgeorge) [reply tests/multi_net/udp_data.py:43]: eg `print_stats(seq)`

Comment (iabdalkader) [reply tests/multi_net/udp_data.py]: Okay fixed.

**Comment by dpgeorge** [comment]
Thanks for updating.

I was testing this and I think that if there are UDP failures then packets from one part of the test could appear as valid for the next part of the test.  The loop over NUM_NEW_SOCKETS should give independent tests with a new socket.  One way to enforce this is to change the port number for each new socket.  I tested using `PORT + i` in the loop (in both client and server) and it seemed to work fine.

What do you think, would that potentially introduce other problems?  If not I think it's a good change to make.

**Comment by iabdalkader** [comment]
> What do you think, would that potentially introduce other problems?

I can't think of any possible way this could fail, and I did some tests, including dropping some first 10 packets intentionally, and seems to working fine, so I made the change.

**Comment by dpgeorge** [comment]
Thanks for updating, merged in 30a022548fcd20bf956789185e1481edb0f0acfc

**Pull request closed without merging**